### PR TITLE
[WIP] Refactor startup scripts

### PR DESCRIPTION
**Pull request by mcsauder**
This PR is a rebase of #9065.  The PR aims to enable future maintainability, reduce and condense hardware specific logic cases, and improve readability of the startup scripts.

Please let me know if you find any unintended breakage due to the significant restructure and please let me know if you have questions or comments.

Thanks!

-Mark

### CODE REVIEW

**Comment by mcsauder** [comment]
As a snapshot, comparing today's PX4/Firmware master branch to this PR with PX4/Firmware master merged, the memory usage is improved by ~8%

Current PX4/Firmware master:
image

PR #8487:
image

**Comment by mcsauder** [comment]
@mhkabir , most echo statements are now removed.  If your preference would be to remove the few remaining in rcS just let me know.  Thanks for your feedback!

-Mark

**Comment by LorenzMeier** [comment]
@mcsauder This looks like great cleanup on a first pass. I will test this on hardware today and see if I can find any corner cases in usage.

**Comment by mcsauder** [comment]
@LorenzMeier , @mhkabir , @dagar , I reviewed my post today and noticed that the RAM comparison was inaccurate because the same processes were not running in both instances.  I've redone that test.  The results are shown below, still positive progress for RAM usage, but not as dramatic, only ~1kb.

I will push up new code to this PR as soon as I have completed flight testing, there was breakage after merging in the past weekend's commits as evidenced by my post above and I will validate with flight testing tomorrow:
image

**Comment by bkueng** [comment]
I would be surprised if RAM consumption considerably decreases since the same processes are supposed to run (it can change a bit due to fragmentation though).

I flashed it on a Pixracer and got some errors:

```
Build datetime: May 22 2018 12:06:15
Build uri: BUILD_URI
Toolchain: GNU GCC, 7.1.0
MFGUID: 3437393931375114002b0033
MCU: STM32F42x, rev. 3
UID: 2B0033:31375114:34373939
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
tune_control [8:100]
INFO  [tune_control] Publishing [standard tune 1
hardfault_log] Fault Log is Armed
INFO  [param] selected parameter default file /fs/mtd_params
WARN  [rgbled] no RGB led on bus #1
WARN  [blinkm] I2C init failed
WARN  [blinkm] init failed
nsh: rgbled_pwm: command not found
rgbled_pwm failed to start
MPU9250 on SPI bus 1 at 4 (1000 KHz)
INFO  [mpu9250] accel cutoff set to 30.00 Hz
INFO  [mpu9250] gyro cutoff set to 80.00 Hz
HMC5883_I2C on I2C bus 1 at 0x1e (bus: 100 KHz, max: 400 KHz)
INFO  [ist8310] no device on bus 1
MS5611_SPI on SPI bus 2 at 3 (20000 KHz)
WARN  [bst] no devices found
WARN  [drivers_board] reset done, 50 ms
WARN  [mpu6000] no device on bus #3 (SPI1)
MPU6000 on SPI bus 1 at 6 (1000 KHz)
INFO  [mpu6000] accel cutoff set to 30.00 Hz
INFO  [mpu6000] gyro cutoff set to 80.00 Hz
WARN  [mpu9250] probe failed! 175
WARN  [mpu9250] no device on bus 4
ERROR [bmi160] driver start failed
WARN  [bmi055] No BMI055 gyro found
ERROR [bmm150] Internal I2C not available
HMC5883_SPI on SPI bus 1 at 5 (11000 KHz)
WARN  [lis3mdl] no device on bus 2
BMP280_I2C on I2C bus 1 at 0x76 (bus: 100 KHz, max: 100 KHz)
WARN  [bmp280] id of your baro is not: 0x58
WARN  [bmp280] bus option number is 2
ERROR [bmp280] driver start failed
adc sample timeout
INFO  [dataman] Unknown restart, data manager file '/fs/microsd/dataman' size is 350216 bytes
INFO  [load_mon] stack check enabled
INFO  [mavlink] mode: Normal, data rate: 20000 B/s on /dev/ttyACM0 @ 921600B
INFO  [mavlink] mode: Normal, data rate: 1200 B/s on /dev/ttyS1 @ 57600B
INFO  [mavlink] mode: OSD, data rate: 1000 B/s on /dev/ttyS2 @ 57600B
ERROR [param] Parameter SYS_USE_IO not found
INFO  [init] Mixer: /etc/mixers/quad_x.main.mix on /dev/pwm_output0
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
ERROR [pwm] can't open /dev/pwm_output1
INFO  [logger] logger started (mode=all)
px4flow [334:100]
WARN  [px4flow] scanning I2C buses for device..
nsh: fi: not valid in this context
Boot has completed
```

**Comment by mcsauder** [comment]
Thanks @bkueng !  I appreciate your feedback.  I was able to get a stock pixracer and duplicate your results.  I've pushed up the commit that corrects the error.

I do see that there are a few additional drivers attempting to be started with this refactor.  I'll do some digging into what cases were removed specific to the pixracer that I don't see in our pixracer based board.  For convenience, here is a screen capture from the startup script outputs from PX4/Firmware master branch as well as from this PR:
image

**Comment by mcsauder** [comment]
The reason the errors @bkueng demonstrated occurred is because of the order of two if statements in rc.interface.  Below is the order that produces the error, and swapping the order corrects the error.

Because the default value of MNT_MODE_OUT is 0, the logic in these two cases could probably use additional attention:
image

**Comment by mcsauder** [comment]
For those interested, here is a screenshot of hte px4fmu-v2_default bloaty output for this PR:  
image

File: ROMFS/px4fmu_common/init.d/rc.io [hunk 1]
```diff
@@ -1,17 +1,93 @@
+#!nsh
 #
-# Init PX4IO interface
+# Script to configure control interface
+#
+#
+# NOTE: environment variable references:
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.io]: I don't think we need to add this everywhere

File: ROMFS/px4fmu_common/init.d/rc.io [hunk 2]
```diff
@@ -1,17 +1,93 @@
...
+set IO_FILE             /etc/extras/
 
 #
-# Adjust PX4IO update rate limit
+# Set USE_IO flag.
 #
-set PX4IO_LIMIT 400
+if param compare SYS_USE_IO 1
+then
+	set USE_IO yes
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.io]: I think we have an ordering problem here: rc.interface sets USE_IO, but rc.io is executed first.
IO_RESENT needs inspection as well.

File: ROMFS/px4fmu_common/init.d/rc.logging [hunk 3]
```diff
@@ -3,13 +3,37 @@
...
+
+#
+# Setup for board specific configurations.
+#
+if ver hwcmp PX4FMU_V5
+then
+	set LOGGER_BUF 64
+fi
+
+if param greater UAVCAN_ENABLE 1
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.logging]: Can you move this into the `else` of `if ver hwcmp PX4FMU_V5`? On V5 we have enough RAM.

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 4]
```diff
@@ -1,119 +1,222 @@
...
+#
+# Setup for generic configurations.
+#
+
+# Configure all I2C buses to 100 KHz as they are all external or slow.
+fmu i2c 1 100000
+fmu i2c 2 100000
+
+# ADC.
+adc start
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: These were not executed on AEROFC_V1 (it uses a different adc for example)

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 5]
```diff
@@ -1,119 +1,222 @@
...
+
+# STMicroelectronics 9-DOF IMU/Magnetometer.
+mpu9250 -s -R 14 start
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: This was not started before unconditionally. What is your reason for adding it?
Similar the ones below.

File: ROMFS/px4fmu_common/init.d/rc.thermal_cal [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+if param compare SYS_CAL_BARO 1
+then
+	set TEMP_CALIB_ARGS "-b"
+	param set SYS_CAL_BARO 0
+fi
+
+#
+# Start the send_event module for temperature calibration
+#
+send_event start
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.thermal_cal]: `send_event` is responsible for more than thermal calibration and thus should be started in rcS

File: ROMFS/px4fmu_common/init.d/rc.interface [hunk 7]
```diff
@@ -9,101 +9,258 @@
...
+set MKBLCTRL_ARGS                       ""
+set SDCARD_MIXERS_PATH                  /fs/microsd/etc/mixers
+set USE_IO                              no
 
-if [ $MIXER != none -a $MIXER != skip ]
+#
+# If mount (gimbal) control is enabled and output mode is AUX, set the aux
+# mixer to mount (override the airframe-specific MIXER_AUX setting).
+#
+if param compare MNT_MODE_OUT 0
```
[the remark this answers is not in the document]
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.interface]: Hi @bkueng , Thanks again for your review and feedback.  I did remove it intentionally, and reviewing, that created the mistake you posted previously.  The corner case didn't reveal itself on our modified configuration.  I've pushed up the fix restoring that logic and it allows the hardware specific checks to take place first.  I appreciate you capturing that.

File: ROMFS/px4fmu_common/init.d/rc.interface [hunk 8]
```diff
@@ -9,101 +9,258 @@
...
+#
+# Set default output mode if not set.
+#
+if [ $OUTPUT_MODE == none ]
+then
+	set OUTPUT_MODE fmu
+fi
+
+# If OUTPUT_MODE == none then something is wrong with setup and we shouldn't try to enable output.
+if [ $OUTPUT_MODE != none ]
```
[the remark this answers is not in the document]
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.interface]: Agreed. I missed additional logic in refactoring.  Another blind spot with limited hardware to test against on the bench. Correction pushed up. Thanks.

File: ROMFS/px4fmu_common/init.d/rc.io (hunk 2 above)
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.io]: Again, thanks for your review @bkueng.  USE_IO is now declared at the top level rcS to allow airframe config files to set/manipulate that parameter for lower level scripts to use as they currently do.  This appears to have silenced one warning that exists in PX4/Firmware master `ERROR [param] Parameter SYS_USE_IO not found`.  

IO_PRESENT logic has been repaired as well.

File: ROMFS/px4fmu_common/init.d/rc.sensors (hunk 5 above)
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: @bkueng, reverting that code does cause breakage in the current construct as pushed up to this PR at the time of writing this comment.  The logic is tangled and I am still attempting to de-tangle and simplify the script.  I will require a short amount of additional time to troubleshoot this set of details and report back.  Thank you again for your feedback!

Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: Ah... @bkueng, those script lines would have occurred as default as long as the hwcmp was not AEROFC_V1 or OMNIBUS_F4SD... this is what made them seem to be acceptable as generic cases.  Please advise me on how you would like this to be handled.

File: ROMFS/px4fmu_common/init.d/rc.logging (hunk 3 above)
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.logging]: Done.  Please let me know if I mis-interpreted your intention. Thanks!

File: ROMFS/px4fmu_common/init.d/rc.sensors (hunk 4 above)
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: Agreed.  Please see comment below.

File: ROMFS/px4fmu_common/init.d/rc.axialracing_ax10_defaults [hunk 9]
```diff
@@ -21,6 +36,10 @@ fi
 set PWM_RATE 50
 
 # Enable servo output on pins 3 and 4 (steering and thrust)
-# but also include 1+2 as they form together one output group
+# but also include 1+2 as they form one output group
 # and need to be set together.
 set PWM_OUT 1234
+
+#
+# Unset all local script variables to free RAM.
```
[the remark this answers is not in the document]
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.axialracing_ax10_defaults]: I shortened this up.  (To explain my rationale, I put the statements at the top and bottom of every script to keep future authors from peppering variables into the body of any of the scripts as we have in the current master branch.)

File: ROMFS/px4fmu_common/init.d/rc.sensors (hunk 5 above)
Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: @bkueng, I went ahead and relocated those driver start lines to each of the other vehicle cases that were utilizing them.  It duplicates those startups across vehicle cases, but I think it is the easiest way to kep future hardware specific logic from growing the number of if/then/else statements as has been the trend.  In this manner, in the future, if another driver is added for any specific board it should be very easy to not affect any other vehicle configuration.

**Comment by mcsauder** [comment]
I've decided to move all script variable declarations and unset calls to rcS as it allows any airframe configuration script to set any of those parameters in the future without disruption of the subordinate scripts.  This also makes housekeeping a very simple effort in the future in ensuring that all script variables are always unset.  I've also been able to break more logic out of rcS into rc.interface, then condense the board specific aspects of that logic.  Testing on a pixracer yields the same number of processes started and everything answers correctly in QGC.  I will continue testing to identify any remaining corner cases as report back.  Thank you all for your time spent reviewing this PR, I appreciate it very much.  If anyone else has feedback please let me know.

**Comment by LorenzMeier** [comment]
Needs a rebase as discussed. Thanks!

**Comment by mcsauder** [comment]
Thanks @LorenzMeier and @bkueng.  I've incorporated #9551 into this PR!

File: ROMFS/px4fmu_common/init.d/rc.interface [hunk 10]
```diff
@@ -179,52 +403,60 @@ then
...
 		fi
 
 		#
-		# Set disarmed, min and max PWM_AUX values
+		# Set disarmed, min and max PWM_AUX values.
 		#
-		if [ $PWM_AUX_DISARMED != none -a $PWM_ACHDIS != none ]
+		if [ $PWM_AUX_DISARMED != none ]
 		then
 			pwm disarmed -c ${PWM_ACHDIS} -p ${PWM_AUX_DISARMED} -d ${OUTPUT_AUX_DEV}
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.interface]: I just tested it on an X2.1 which has an IO, and this is executed as:
```
pwm disarmed -c  -p p:PWM_AUX_DISARMED -d /dev/pwm_output1
```
So `PWM_ACHDIS` is not set.

Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.interface]: @bkueng, I've pushed up a commit that deprecates `PWM_ACHDIS` and I certainly feel it improves/simplifies the logic flow in the script and config files.  Could you give me feedback given your test hardware setup at any time it is convenient?  Thanks!

Comment (bkueng) [reply ROMFS/px4fmu_common/init.d/rc.interface]: The change looks good

**Comment by bkueng** [comment]
Tested again, here is the boot output comparison (right=this PR):
screenshot_20180608_095315
Looks good, except that now another mavlink instance is started on /dev/ttyS0

**Comment by mcsauder** [comment]
@bkueng,  I corrected the mistake that caused your extra mavlink instance.  That commit is pushed up.

Thank you again for all of your testing and feedback!  I appreciate it a lot.

-Mark

File: ROMFS/px4fmu_common/init.d/rc.mavlink [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+	frsky_telemetry start -d /dev/ttyS6
+
+	# Setup MAVLink on Wifi (ESP8266 port).
+	set SERIAL_DEVICE_0_ARGS "-r 20000"
+	set SERIAL_DEVICE_0_BAUD "921600"
+
+	# Use ttyS1 for MAVLink on FMUv4 in addition to ttyS0 (debug).
+	set SERIAL_DEVICE_1_ARGS "-r 1200"
+
+	# Start MAVLink on Wifi (ESP8266 port).
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.mavlink]: There is something wrong here: we have this twice now, also for `SERIAL_DEVICE_0_BAUD` right above (which specifies the baudrate and args correctly).

Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.mavlink]: @bkueng Agreed.  I have dug into this and discovered additional cases that I am untangling now. Mavlink startups were the genesis of this PR, so it is fitting that they revisit me at this point in the work.  More work to follow.

**Comment by bkueng** [comment]
Now that the release is out we can come back to this.

I tested that HIL still works.

@mcsauder can you rebase one last time and fix the pixracer mavlink wifi?

**Comment by mcsauder** [comment]
Hi @bkueng,

I had to do a fair bit of work to untangle two cases that were goofed up by my rc.mavlink work in this refactor.  My testing shows correct mavlink stream behavior on the hardware I have access to.  If you find otherwise in your testing please let me know.

In short, I've grouped the hardware specific cases at the top of rc.mavlink and put in place a framework that will allow users to migrate away from using SYS_COMPANION and I've created a pattern for starting additional UART, USB, or generic DEV (device) mavlink streams that can be utilized by the airframe config files.

Unfortunately, it appears that adding the extra MAVLINK_UART_X instances has grown the binary image size and it is failing again for px4fmu-v2, so I probably need to make one more evaluation/commit.  I'll get that attended to this weekend.

Please let me know if you turn up any cases I've missed in your testing.  Thanks for all of your feedback!

-Mark

File: ROMFS/px4fmu_common/init.d/rc.mavlink [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+fi
+
+if param compare SYS_COMPANION 419200
+then
+	# TODO: this should now be relocated with a board specific configuration above.
+	set MAVLINK_DEV_0       "/dev/iridium"
+	set MAVLINK_DEV_0_ARGS  "-m iridium -r 10"
+	set MAVLINK_DEV_0_BAUD  "19200"
+
+	iridiumsbd start -d /dev/ttyS2
```
Comment (bkueng) [inline ROMFS/px4fmu_common/init.d/rc.mavlink]: This will conflict with the default of `MAVLINK_UART_2`. I think we should be using `MAVLINK_UART_2` here and drop `MAVLINK_DEV_0`. What do you think?

Also the USB mavlink instance does IMO not need to be configurable. This will reduce Flash usage further. You can also revert 2a3d66cf457e9b7879276bd01be7936876b56147.

Comment (mcsauder) [reply ROMFS/px4fmu_common/init.d/rc.mavlink]: Thanks @bkueng,

I fixed up the MAVLINK_UART_2 conflict you identified, (thanks for spotting that), and I've removed the MAVLINK_USB_0 and MAVLINK_DEV_0 instances and reverted the commit you pointed out to save on flash.

I've reverted both the hardware specific cases at the top of rc.mavlink to simply start their respective mavlink instances to save on flash, however it still isn't enough.

Another ~2k of flash can be reclaimed by deprecating the usage of the `SYS_COMPANION` in rc.mavlink.  Airframe config files in this PR have been updated to allow this.  Let me know if this is acceptable to you.

Thanks!

-Mark

Comment (bkueng) [reply ROMFS/px4fmu_common/init.d/rc.mavlink]: Thanks for the update.

> Another ~2k of flash can be reclaimed by deprecating the usage of the SYS_COMPANION in rc.mavlink. Airframe config files in this PR have been updated to allow this. Let me know if this is acceptable to you.

My answer is still the same as in https://github.com/PX4/Firmware/pull/9065#issuecomment-390106087. I'd like to push for something like https://github.com/PX4/Firmware/pull/9615, then all UARTs are fully configurable via QGC with minimal flash overhead. The airframe configs can then also override the defaults.

**Comment by LorenzMeier** [comment]
Awesome, thanks!

**Comment by mcsauder** [comment]
Work originally intended by this PR has been all wrapped up through the other merge requests!  Thanks everyone!
